File: packages/domain/src/query-packs/classification.ts
/**
 * Signal-strength classification for query-pack matches (BB-038).
 * Weak signals may produce candidates only — never independent promotion.
 */
import type {
  MatchOutcome,
  QueryTerm,
  SignalStrength,
  TermClass,
} from './types.js';

const STRONG_CLASSES: ReadonlySet<TermClass> = new Set(['positive', 'historical', 'modern']);
const SUPPORTING_CLASSES: ReadonlySet<TermClass> = new Set(['geographic', 'alias', 'source_specific']);
const WEAK_ONLY_CLASSES: ReadonlySet<TermClass> = new Set(['negative', 'alias', 'geographic']);

export type ClassifySignalInput = {
  readonly matchedTerms: readonly QueryTerm[];
};

export type ClassifySignalResult = {
  readonly strength: SignalStrength;
  readonly outcome: MatchOutcome;
  readonly matchedClasses: readonly TermClass[];
  readonly reasons: readonly string[];
};

function uniqueClasses(terms: readonly QueryTerm[]): TermClass[] {
  return [...new Set(terms.map((term) => term.termClass))];
}

function hasAnyClass(classes: readonly TermClass[], targets: ReadonlySet<TermClass>): boolean {
  return classes.some((value) => targets.has(value));
}

/**
 * Classify matched terms into strong / medium / weak signal strength.
 * - strong: positive paired with historical/modern, or positive + geographic
 * - medium: positive alone, or alias/source_specific with supporting context
 * - weak: negative-only, alias-only, or geographic-only matches
 */
export function classifySignalStrength(input: ClassifySignalInput): ClassifySignalResult {
  const matchedClasses = uniqueClasses(input.matchedTerms);
  const reasons: string[] = [];

  if (input.matchedTerms.length === 0) {
    return {
      strength: 'weak',
      outcome: 'candidate_only',
      matchedClasses,
      reasons: ['no_terms_matched'],
    };
  }

  const hasPositive = matchedClasses.includes('positive');
  const hasHistoricalOrModern =
    matchedClasses.includes('historical') || matchedClasses.includes('modern');
  const hasGeographic = matchedClasses.includes('geographic');
  const hasAlias = matchedClasses.includes('alias');
  const hasSourceSpecific = matchedClasses.includes('source_specific');
  const hasNegative = matchedClasses.includes('negative');

  if (hasPositive && hasHistoricalOrModern) {
    reasons.push('positive_with_period_term');
    return { strength: 'strong', outcome: 'promotable', matchedClasses, reasons };
  }

  if (hasPositive && hasGeographic) {
    reasons.push('positive_with_geographic');
    return { strength: 'strong', outcome: 'promotable', matchedClasses, reasons };
  }

  if (hasPositive && (hasAlias || hasSourceSpecific)) {
    reasons.push('positive_with_alias_or_source');
    return { strength: 'medium', outcome: 'promotable', matchedClasses, reasons };
  }

  if (hasPositive) {
    reasons.push('positive_only');
    return { strength: 'medium', outcome: 'promotable', matchedClasses, reasons };
  }

  if (hasHistoricalOrModern && !hasPositive) {
    reasons.push('period_term_without_positive');
    return { strength: 'weak', outcome: 'candidate_only', matchedClasses, reasons };
  }

  if (hasNegative && matchedClasses.length === 1) {
    reasons.push('negative_only');
    return { strength: 'weak', outcome: 'candidate_only', matchedClasses, reasons };
  }

  if (hasAlias && matchedClasses.every((value) => WEAK_ONLY_CLASSES.has(value))) {
    reasons.push('alias_without_positive');
    return { strength: 'weak', outcome: 'candidate_only', matchedClasses, reasons };
  }

  if (hasGeographic && !hasAnyClass(matchedClasses, STRONG_CLASSES)) {
    reasons.push('geographic_without_positive');
    return { strength: 'weak', outcome: 'candidate_only', matchedClasses, reasons };
  }

  if (hasSourceSpecific && hasAnyClass(matchedClasses, SUPPORTING_CLASSES) && !hasPositive) {
    reasons.push('source_specific_without_positive');
    return { strength: 'weak', outcome: 'candidate_only', matchedClasses, reasons };
  }

  reasons.push('residual_medium');
  return { strength: 'medium', outcome: 'promotable', matchedClasses, reasons };
}

export function outcomeForSignalStrength(strength: SignalStrength): MatchOutcome {
  return strength === 'weak' ? 'candidate_only' : 'promotable';
}

export function assertMayPromoteBeyondCandidate(result: ClassifySignalResult): void {
  if (result.outcome === 'candidate_only') {
    throw new Error(
      `Weak signal (${result.strength}) produces candidates only: ${result.reasons.join(', ')}`,
    );
  }
}

export function mayPromoteBeyondCandidate(result: ClassifySignalResult): boolean {
  return result.outcome === 'promotable';
}

File: workers/research/src/black_book_research/query_packs/classification.py
"""Signal-strength classification for query-pack matches (BB-038)."""

from __future__ import annotations

from .types import ClassifySignalResult, MatchOutcome, QueryTerm, SignalStrength, TermClass

STRONG_CLASSES: frozenset[TermClass] = frozenset({"positive", "historical", "modern"})
SUPPORTING_CLASSES: frozenset[TermClass] = frozenset({"geographic", "alias", "source_specific"})
WEAK_ONLY_CLASSES: frozenset[TermClass] = frozenset({"negative", "alias", "geographic"})


def _unique_classes(terms: tuple[QueryTerm, ...]) -> tuple[TermClass, ...]:
    seen: list[TermClass] = []
    for term in terms:
        if term.term_class not in seen:
            seen.append(term.term_class)
    return tuple(seen)


def classify_signal_strength(*, matched_terms: tuple[QueryTerm, ...]) -> ClassifySignalResult:
    matched_classes = _unique_classes(matched_terms)
    reasons: list[str] = []

    if not matched_terms:
        return ClassifySignalResult(
            strength="weak",
            outcome="candidate_only",
            matched_classes=matched_classes,
            reasons=("no_terms_matched",),
        )

    has_positive = "positive" in matched_classes
    has_historical_or_modern = "historical" in matched_classes or "modern" in matched_classes
    has_geographic = "geographic" in matched_classes
    has_alias = "alias" in matched_classes
    has_source_specific = "source_specific" in matched_classes
    has_negative = "negative" in matched_classes

    if has_positive and has_historical_or_modern:
        reasons.append("positive_with_period_term")
        return ClassifySignalResult("strong", "promotable", matched_classes, tuple(reasons))

    if has_positive and has_geographic:
        reasons.append("positive_with_geographic")
        return ClassifySignalResult("strong", "promotable", matched_classes, tuple(reasons))

    if has_positive and (has_alias or has_source_specific):
        reasons.append("positive_with_alias_or_source")
        return ClassifySignalResult("medium", "promotable", matched_classes, tuple(reasons))

    if has_positive:
        reasons.append("positive_only")
        return ClassifySignalResult("medium", "promotable", matched_classes, tuple(reasons))

    if has_historical_or_modern and not has_positive:
        reasons.append("period_term_without_positive")
        return ClassifySignalResult("weak", "candidate_only", matched_classes, tuple(reasons))

    if has_negative and len(matched_classes) == 1:
        reasons.append("negative_only")
        return ClassifySignalResult("weak", "candidate_only", matched_classes, tuple(reasons))

    if has_alias and all(value in WEAK_ONLY_CLASSES for value in matched_classes):
        reasons.append("alias_without_positive")
        return ClassifySignalResult("weak", "candidate_only", matched_classes, tuple(reasons))

    if has_geographic and not any(value in STRONG_CLASSES for value in matched_classes):
        reasons.append("geographic_without_positive")
        return ClassifySignalResult("weak", "candidate_only", matched_classes, tuple(reasons))

    if has_source_specific and any(value in SUPPORTING_CLASSES for value in matched_classes) and not has_positive:
        reasons.append("source_specific_without_positive")
        return ClassifySignalResult("weak", "candidate_only", matched_classes, tuple(reasons))

    reasons.append("residual_medium")
    return ClassifySignalResult("medium", "promotable", matched_classes, tuple(reasons))


def outcome_for_signal_strength(strength: SignalStrength) -> MatchOutcome:
    return "candidate_only" if strength == "weak" else "promotable"


def assert_may_promote_beyond_candidate(result: ClassifySignalResult) -> None:
    if result.outcome == "candidate_only":
        joined = ", ".join(result.reasons)
        raise ValueError(f"Weak signal ({result.strength}) produces candidates only: {joined}")


def may_promote_beyond_candidate(result: ClassifySignalResult) -> bool:
    return result.outcome == "promotable"
